kythe/java/com/google/devtools/kythe/platform/java/JavacOptionsUtils.java

```diff
@@ -61,177 +67,209 @@
-          continue outer;
+
+  /** A useful container for modifying javac commandline arguments, in the style of a builder. */
+  public static class ModifiableOptions {
+    private List<String> internal = new ArrayList<>();
+
+    private ModifiableOptions() {}
+
+    public static ModifiableOptions of() {
+      return of(ImmutableList.of());
+    }
+
+    public static ModifiableOptions of(Iterable<String> options) {
+      ModifiableOptions o = new ModifiableOptions();
+      Iterables.addAll(o.internal, options);
+      return o;
+    }
+
+    public ImmutableList<String> build() {
+      return ImmutableList.copyOf(internal);
+    }
+
+    public ModifiableOptions add(String option) {
+      internal.add(option);
+      return this;
+    }
+
+    /** Removes unsupported javac compiler options. */
+    public ModifiableOptions removeUnsupportedOptions() {
+      List<String> previous = internal;
+      internal = new ArrayList<>();
+      ImmutableList<OptionChecker> optionCheckers =
+          ImmutableList.of(JavacTool.create(), new JavacFileManager(new Context(), false, UTF_8));
+      PeekingIterator<String> it =
+          Iterators.peekingIterator(
+              Iterators.filter(
+                  previous.iterator(),
+                  Predicates.not(o -> o.startsWith("-Xlint") || o.startsWith("-Werror"))));
+      outer:
+      while (it.hasNext()) {
+        for (OptionChecker optionChecker : optionCheckers) {
+          int arity = optionChecker.isSupportedOption(it.peek());
+          if (arity > 0 && it.peek().indexOf(':') != -1) {
+            // For "conjoined" flags (e.g., -flag:arg) we want to consume just command-line option.
+            arity = 0;
+          }
+          if (arity != -1) {
+            Iterators.addAll(internal, Iterators.limit(it, arity + 1));
+            continue outer;
+          }
         }
+        it.next();
       }
-      it.next();
+      return this;
     }
-    return options;
-  }
 
-  /** Removes the given {@link Option}s (and their arguments) from {@code options}. */
-  public static List<String> removeOptions(List<String> options, Set<Option> opts) {
-    for (int i = 0; i < options.size(); ) {
-      String opt = options.get(i);
-      boolean matched = false;
-      for (Option o : opts) {
-        if (o.matches(opt)) {
-          matched = true;
-          options.subList(i, i + (o.hasArg() ? 2 : 1)).clear();
-          break;
+    /** Removes the given {@link Option}s (and their arguments) from the builder. */
+    public ModifiableOptions removeOptions(Set<Option> opts) {
+      List<String> original = internal;
+      internal = new ArrayList<>(original.size());
+      for (int i = 0; i < original.size(); i++) {
+        String opt = original.get(i);
+        boolean matched = false;
+        for (Option o : opts) {
+          if (o.matches(opt)) {
+            matched = true;
+            if (o.hasArg()) {
+              // Skip the argument too.
+              i++;
```

Comment:
This is nearly identical (semantically, if not syntactically) with `removeUnsupported` and should probably be unified.

Reply:
My favorite pattern for dealing with this comes from the JavaFileManager class and is essentially the interface:

interface OptionHandler {
Iterator<String> handleOption(String value, Iterator<String> remaining);
}

// remove
(opt, rest) -> rest
// w/ value
(opt, rest) -> { rest.next(); return rest; }
// keep
(opt, rest) -> Iterator.concat(ImmutableList.of(opt), rest);

etc.

Reply:
ehh, maybe just the one loop can do it without arbitrary handler.

But, similar logic for different handling of what to do on match / not-match / positionally-matched makes sense still imo.